### PR TITLE
cluster/state: refactor to proto serialisation

### PR DESCRIPTION
**Pull request by corverroos**
Refactors cluster state serialisation from JSON to proto and from hashing from SSZ to simple SHA256 hashing combined with anypb.

Protos provide the following benefits:
- This removes complexity of SSZ hashing
- Removes the complexity of JSON serialisation
- Schema first protobuf definitions 

category: refactor
ticket: none

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -53,8 +53,8 @@ github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym
 github.com/Microsoft/go-winio v0.6.1 h1:9/kr64B9VUZrLm5YYwbGtUJnMgqWVOdUAXu6Migciow=
 github.com/Microsoft/go-winio v0.6.1/go.mod h1:LRdKpFKfdobln8UmuiYcKPot9D2v6svN5+sAH+4kjUM=
 github.com/anmitsu/go-shlex v0.0.0-20161002113705-648efa622239/go.mod h1:2FmKhYUyUczH0OGQWaF5ceTx0UBShxjsH6f8oGKYe2c=
-github.com/attestantio/go-eth2-client v0.16.6 h1:Vpn54vQig48YgqyC2rAXUpUzxgjtZsSNs9HcNEL0oMg=
-github.com/attestantio/go-eth2-client v0.16.6/go.mod h1:lgxKsjRSxQnHSWxSbtZLGYlFkE3vFM4L+DK4bXqAxAU=
+github.com/attestantio/go-eth2-client v0.17.0 h1:Rvn/tmLHRRztoS2c/6AmsslGucyytWMvnQlAUz4EvYY=
+github.com/attestantio/go-eth2-client v0.17.0/go.mod h1:N+BNIxaHmul44d0tTjwFrUJU/g6MIblFP67oagm1Lwo=
```
Comment (corverroos) [inline go.sum:57]: will revert

**Review by gsora — APPROVED** [review]
Nitpicky comment, A+ regardless

File: cluster/state/cluster.go [hunk 2]
```diff
@@ -108,175 +107,17 @@ func (c Cluster) NodeIdx(pID peer.ID) (cluster.NodeIdx, error) {
...
-	PubKey              []byte      `ssz:"ByteList[256]"`
-	PubShares           []sszPubkey `ssz:"CompositeList[65536]"`
-	FeeRecipientAddress []byte      `ssz:"Bytes20"`
-	WithdrawalAddress   []byte      `ssz:"Bytes20"`
-}
-
-type sszPubkey struct {
-	Pubkey []byte `ssz:"ByteList[256]"`
+// ValidatorPublicKey returns the valIdx'th validator BLS group public key.
+func (c Cluster) ValidatorPublicKey(valIdx int) (tbls.PublicKey, error) {
```
Comment (xenowits) [inline cluster/state/cluster.go:111]: nit: `valIdx` is confusing to me as it refers to `Validator Index` on beacon chain which it is not. I suppose renaming it to `indx` would be fine

File: cluster/statepb/v1/state.proto [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  bytes                        parent = 1; // Parent is the hash of the parent mutation. It must be 32 bytes.
+  string                         type = 2; // Type is the type of mutation.
+  google.protobuf.Timestamp timestamp = 3; // Timestamp is the time of the mutation.
+  google.protobuf.Any            data = 4; // Data is the data of the mutation. Must be non-nil.
+}
+
+// SignedMutation is a mutation signed by a signer.
+message SignedMutation {
+  Mutation mutation = 1; // Mutation is the mutation.
+  bytes      signer = 2; // Signer is the identify (public key) of the signer.
```
Comment (xenowits) [inline cluster/statepb/v1/state.proto:21]:
```suggestion
  bytes      signer = 2; // Signer is the identity (public key) of the signer.
```

**Review by xenowits — APPROVED** [review]

File: cluster/distvalidator.go [hunk 4]
```diff
@@ -39,6 +45,17 @@ func (v DistValidator) PublicShare(peerIdx int) (tbls.PublicKey, error) {
 	return tblsconv.PubkeyFromBytes(v.PubShares[peerIdx])
 }
 
+// HasRegistration returns a true if the validator has a valid registration.
+func (v DistValidator) HasRegistration() bool {
```
[the remark this answers is not in the document]
Comment (corverroos) [reply cluster/distvalidator.go:49]: the idea was the detect a zero registration vs a invalid registration, but I'll move the zero checks here rather and then call this in `Eth2Registration`

File: cluster/statepb/v1/state.proto (hunk 3 above)
Comment (corverroos) [reply cluster/statepb/v1/state.proto]: done

**Pull request merged**